### ISSUE TITLE
Ignore promises in S3735 `void-use`

### ISSUE THREAD

**Issue by vilchik-elena** [SonarSource/SonarJS#2629]
eslint@typescript-eslint/no-floating-promises is reporting when promise is used not being used and suggests to add `void` to be explicit that value is ignore.
```
  const exampleFunction = () => { return new Promise(() => {});};
  void exampleFunction();
```

For such code rule S3735 will report that `void` should not be used. We should improve S3735 to not report when void argument has promise type.
 
Reported in https://community.sonarsource.com/t/typescript-void-should-not-be-used-triggers-wrongly/41893

**Comment by sebastien-savalle** [comment]
This issue is really annoying. Do you have any news ?

**Comment by vglaeser** [comment]
It seems that this issue has been "unfixed" in the meantime (using SonarQube@10.1).  
In my case, it shows a Code smell for the following lines of code in an Angular project (I removed some injections and additional calls in the constructor's body):
```
constructor(
    private router: Router,
  ) {
    this.router.navigate(['']);
  }

**Comment by gitgdako** [comment]
We are facing the same issue in several different cases. We can provide some code samples if necessary. Shall we create a new issue?

**Comment by JustinHorn** [comment]
I also just encounted this issue. I guess this one needs to get reopened.

**Comment by nirtamir2** [comment]
```tsx
  void window.navigator.share({
          title,
          text,
          url: `${window.location.origin}${path}`,
        });
      }}
```

False positive

**Comment by jiikoosol** [comment]
We are are having the exact same issue in our project

**Comment by mmospanenko** [comment]
yes, this is mistake, voiding promises is documented feature. Why is this issue closed without resolution?

**Comment by AnMW** [comment]
Hi, I'm also experiencing this issue right now, so I wanted to comment to give this a bump :)

**Comment by ericmorand-sonarsource** [comment]
Hi,

Thanks a lot for the reports. I can confirm that the _void is allowed with Promise-like objects_ is not working without type information - e.g. when a code is written with JavaScript. I created a Jira ticket to track the discussion about this issue: the documentation states that the rule applies to both TypeScript and JavaScript so either the documentation should be updated to highlight the fact that the exception only applies to TypeScript, or the rule should be fixed so that the exception also applies to JavaScript.

Allow me to give my personal opinion: `window.navigator.share` can reject, and this rejection should handled. I understand that, in a browser environment, unhandled rejections are likely to not make the browser process crash, but I believe that no issue should be left unnoticed and that, if `window.navigator.share` rejects, action should be taken.
